### PR TITLE
Core: Add docs nodes to story index

### PR DESCRIPTION
**Pull request by shilman**
Issue: N/A

## What I did

- [x] Add a `storyImports` field to entries for docs2-style story embedding
- [x] Bump the index version to 4
- [x] Rename `stories` to `entries`
- [x] Add a nullable `type` to index entries for `docs` entries
- [x] Add a couple v3 export points for legacy indices
- [x] Add `.docs.mdx` handling

Possibly temporarily:
- [x] Make docs-only stories of type `docs` for testing purposes

Still to do:
- [x] Test all this stuff better

## How to test

```
yarn bootstrap --core
cd examples/react-ts
yarn storybook
open http://localhost:6006/stories.json
open http://localhost:6006/index.json
```

Alternatively, I wrote this little script `yarn ts-node ./lib/core-server/src/utils/testIndex.ts` until we get `jest` working, for fast-loop testing:

```ts
import path from 'path';
import { normalizeStoriesEntry } from '@storybook/core-common';
import type { NormalizedStoriesSpecifier } from '@storybook/core-common';
import { StoryIndexGenerator } from './StoryIndexGenerator';

const options = {
  configDir: path.join(__dirname, '__mockdata__'),
  workingDir: path.join(__dirname, '__mockdata__'),
  storiesV2Compatibility: false,
  storyStoreV7: true,
};

const main = async () => {
  const storiesSpecifier: NormalizedStoriesSpecifier = normalizeStoriesEntry(
    './src/A.stories.(ts|js|jsx)',
    options
  );
  const docsSpecifier: NormalizedStoriesSpecifier = normalizeStoriesEntry(
    './src/**/MetaOf.docs.mdx',
    options
  );

  const generator = new StoryIndexGenerator([storiesSpecifier, docsSpecifier], options);
  await generator.initialize();
  const index = await generator.getIndex();

  console.log(JSON.stringify(index, null, 2));
};

main()
  .then(() => console.log('done'))
  .catch((err) => console.log({ err }));
```

### CODE REVIEW

File: lib/client-api/src/ClientApi.ts [hunk 1]
```diff
@@ -396,10 +397,12 @@ Read more here: https://github.com/storybookjs/storybook/blob/master/MIGRATION.m
   };
 
   getStorybook = (): GetStorybookKind<TFramework>[] => {
-    const { stories } = this.storyStore.storyIndex;
+    const { entries } = this.storyStore.storyIndex;
 
     const kinds: Record<ComponentTitle, GetStorybookKind<TFramework>> = {};
-    Object.entries(stories).forEach(([storyId, { title, name, importPath }]) => {
+    Object.entries(entries).forEach(([storyId, { title, name, importPath, type }]) => {
+      if (type && type !== 'story') return;
```
Comment (tmeasday) [inline lib/client-api/src/ClientApi.ts:404]: Should we normalize it when we load it into the store so we don't have to do these null checks all over the place?

File: lib/core-server/src/utils/StoryIndexGenerator.test.ts [hunk 2]
```diff
@@ -62,21 +63,23 @@ describe('StoryIndexGenerator', () => {
 
         expect(await generator.getIndex()).toMatchInlineSnapshot(`
           Object {
-            "stories": Object {
+            "entries": Object {
               "nested-button--story-one": Object {
                 "id": "nested-button--story-one",
                 "importPath": "./src/nested/Button.stories.ts",
                 "name": "Story One",
+                "storiesImports": Array [],
```
Comment (tmeasday) [inline lib/core-server/src/utils/StoryIndexGenerator.test.ts:71]: I think this should be unset also for `type: 'story'` - or was there some reason we needed it?

File: lib/core-server/src/utils/stories-json.ts [hunk 3]
```diff
@@ -80,4 +81,31 @@ export async function useStoriesJson(
...
+    } catch (err) {
+      res.status(500);
+      res.send(err.message);
+    }
+  });
 }
+
+const isStory = (entry: StoryIndexEntry) => !entry.type || entry.type === 'story';
+
+export const convertToIndexV3 = (index: StoryIndex): StoryIndexV3 => {
```
Comment (tmeasday) [inline lib/core-server/src/utils/stories-json.ts:101]: Add a test for this?

File: lib/core-server/src/utils/stories-json.ts [hunk 4]
```diff
@@ -80,4 +81,31 @@ export async function useStoriesJson(
       res.send(err.message);
     }
   });
+
+  router.use('/stories.json', async (req: Request, res: Response) => {
```
Comment (tmeasday) [inline lib/core-server/src/utils/stories-json.ts:85]: Are we going to write this on build too? That seems more crucial (composition with v6 storybooks).

File: lib/store/src/StoryStore.ts [hunk 5]
```diff
@@ -262,24 +263,28 @@ export class StoryStore<TFramework extends AnyFramework> {
...
-            fileName: this.storyIndex.stories[story.id].importPath,
-          },
-        }),
-      })
+      (story) => {
+        const { importPath, storiesImports } = this.storyIndex.entries[story.id];
+        return {
+          ...pick(story, ['id', 'name', 'title']),
+          importPath,
+          storiesImports,
```
Comment (tmeasday) [inline lib/store/src/StoryStore.ts:277]: Not part of the v3 spec, right?

**Review by tmeasday — COMMENTED** [review]
Some comments

File: lib/api/src/lib/stories.ts [hunk 6]
```diff
@@ -114,15 +114,22 @@ export interface StoriesRaw {
 }
 
 type Path = string;
-export interface StoryIndexStory {
+export interface StoryIndexEntry {
   id: StoryId;
   name: StoryName;
   title: ComponentTitle;
   importPath: Path;
+  type?: 'story' | 'docs';
```
Comment (tmeasday) [inline lib/api/src/lib/stories.ts:122]: Where's `storiesImports`?

File: lib/core-server/src/utils/StoryIndexGenerator.ts [hunk 7]
```diff
@@ -56,45 +66,145 @@ export class StoryIndexGenerator {
...
-  async ensureExtracted(): Promise<StoryIndex['stories'][]> {
-    return (
-      await Promise.all(
-        this.specifiers.map(async (specifier) => {
-          const entry = this.storyIndexEntries.get(specifier);
-          return Promise.all(
-            Object.keys(entry).map(
-              async (absolutePath) =>
-                entry[absolutePath] || this.extractStories(specifier, absolutePath)
+  async updateExtracted(
```
Comment (tmeasday) [inline lib/core-server/src/utils/StoryIndexGenerator.ts:77]: Can you add a comment

File: lib/core-server/src/utils/StoryIndexGenerator.ts [hunk 8]
```diff
@@ -56,45 +66,145 @@ export class StoryIndexGenerator {
...
+
+  isDocsMdx(absolutePath: Path) {
+    return /\.docs\.mdx$/i.test(absolutePath);
+  }
+
+  async ensureExtracted(): Promise<StoryIndexEntry[]> {
+    await this.updateExtracted(async (specifier, absolutePath) =>
+      this.isDocsMdx(absolutePath) ? false : this.extractStories(specifier, absolutePath)
+    );
+    // process docs in a second pass
```
Comment (tmeasday) [inline lib/core-server/src/utils/StoryIndexGenerator.ts:100]: Explain why?

File: lib/core-server/src/utils/StoryIndexGenerator.ts [hunk 9]
```diff
@@ -56,45 +66,145 @@ export class StoryIndexGenerator {
...
+  }
+
+  async extractDocs(specifier: NormalizedStoriesSpecifier, absolutePath: Path) {
+    const relativePath = path.relative(this.options.workingDir, absolutePath);
+    try {
+      const normalizedPath = normalizeStoryPath(relativePath);
+      const importPath = slash(normalizedPath);
+      const defaultTitle = autoTitleFromSpecifier(importPath, specifier);
+
+      // eslint-disable-next-line global-require
```
Comment (tmeasday) [inline lib/core-server/src/utils/StoryIndexGenerator.ts:122]: Explain why?

File: lib/core-server/src/utils/StoryIndexGenerator.ts [hunk 10]
```diff
@@ -56,45 +66,145 @@ export class StoryIndexGenerator {
...
-          );
-        })
-      )
-    ).flat();
+          : otherImport;
+
+      const absoluteImports = (result.imports as string[]).map(makeAbsolute);
+      const absoluteOf = result.of && makeAbsolute(result.of);
+
+      let ofTitle: string;
```
Comment (tmeasday) [inline lib/core-server/src/utils/StoryIndexGenerator.ts:141]: Explain what you are doing here?

File: lib/core-server/src/utils/StoryIndexGenerator.ts [hunk 11]
```diff
@@ -56,45 +66,145 @@ export class StoryIndexGenerator {
...
+              ofTitle = cacheEntry.entries[0].title;
+            }
+            dependencies.push(cacheEntry);
+          } else {
+            throw new Error(`Unexpected dependency: ${cacheEntry}`);
+          }
+        });
+      });
+
+      dependencies.forEach((dep) => {
```
Comment (tmeasday) [inline lib/core-server/src/utils/StoryIndexGenerator.ts:160]:
```suggestion
      // Track that we depend on this for easy invalidation later.
      dependencies.forEach((dep) => {
```

File: lib/core-server/src/utils/StoryIndexGenerator.ts [hunk 12]
```diff
@@ -56,45 +66,145 @@ export class StoryIndexGenerator {
...
       const csf = (await readCsfOrMdx(absolutePath, { defaultTitle })).parse();
-      csf.stories.forEach(({ id, name }) => {
-        fileStories[id] = {
+      const storiesImports = await Promise.all(
+        csf.imports.map(async (otherImport) =>
+          otherImport.startsWith('.')
+            ? slash(normalizeStoryPath(path.join(path.dirname(normalizedPath), otherImport)))
+            : otherImport
+        )
+      );
```
Comment (tmeasday) [inline lib/core-server/src/utils/StoryIndexGenerator.ts:197]: Is this for `docsOnly`? I think maybe we should put the whole thing in a `if()` block for clarity.

File: lib/core-server/src/utils/StoryIndexGenerator.ts [hunk 13]
```diff
@@ -165,21 +273,31 @@ export class StoryIndexGenerator {
...
   invalidate(specifier: NormalizedStoriesSpecifier, importPath: Path, removed: boolean) {
     const absolutePath = slash(path.resolve(this.options.workingDir, importPath));
-    const pathToEntries = this.storyIndexEntries.get(specifier);
+    const cache = this.specifierToCache.get(specifier);
+
+    const cacheEntry = cache[absolutePath];
+    let dependents = [];
+    if (cacheEntry && cacheEntry.type === 'stories') {
+      dependents = cacheEntry.dependents;
+      // FIXME: might be in another cache
```
Comment (tmeasday) [inline lib/core-server/src/utils/StoryIndexGenerator.ts:291]: FIXME

**Review by tmeasday — COMMENTED** [review]
We definitely need some tests for the `StoryIndexGenerator` changes, esp around invalidation, but I guess you are blocked on that?

File: lib/core-server/src/utils/stories-json.ts (hunk 4 above)
Comment (shilman) [reply lib/core-server/src/utils/stories-json.ts:85]: We already do

**Pull request closed without merging**